### PR TITLE
Avoid persisting unvalidated plugin refs

### PR DESCRIPTION
**Pull request by wemrysi**
Fixes a bug in all the plugin managers that allowed unvalidated plugin refs to be saved when replacing an existing ref.

[ch10943]

### CODE REVIEW

File: impl/src/test/scala/quasar/impl/datasources/DefaultDatasourcesSpec.scala [hunk 1]
```diff
@@ -310,6 +310,32 @@ object DefaultDatasourcesSpec extends DatasourcesSpec[IO, Stream[IO, ?], String,
...
+          c <- dses.replaceDatasource(i, a2)
+          r2 <- refs.lookup(i)
+          d <- dses.quasarDatasourceOf(i)
+          started <- starts.get
+          ended <- shuts.get
+          _ <- finalize
+        } yield {
+          // once for initial and another for restart after failed replace
+          started === List(i, i)
+          ended === List(i)
```
Comment (alissapajer) [inline impl/src/test/scala/quasar/impl/datasources/DefaultDatasourcesSpec.scala:333]: Do you mean to use `must_=` here (or whatever the scalaz/cats variant is)?

File: impl/src/main/scala/quasar/impl/datasources/DefaultDatasources.scala [hunk 2]
```diff
@@ -83,93 +84,70 @@ private[impl] final class DefaultDatasources[
...
-          }
-        } yield res
+
+        // The last value we knew about has since been externally updated,
+        // but our update replaces it, so we ignore the new value and
+        // replace the old.
+        case Updated(_, old) =>
+          doReplace(old)
+
+        case exists: Exists[DatasourceRef[C]] =>
```
Comment (alissapajer) [inline impl/src/main/scala/quasar/impl/datasources/DefaultDatasources.scala:115]: `Updated` is also an `Exists`. Are we gaining anything from having `Exists` or could we just match on the remaining two cases (`Inserted` and `Preserved`) here?

Comment (alissapajer) [reply impl/src/main/scala/quasar/impl/datasources/DefaultDatasources.scala:115]: Also, do we need both `Inserted` and `Preserved` cases anymore, now that they are treated the same here?

Comment (wemrysi) [reply impl/src/main/scala/quasar/impl/datasources/DefaultDatasources.scala:115]: I think we could unify `Inserted` and `Preserved` into `Value` or `Present` and do away with `Exists`.

File: impl/src/test/scala/quasar/impl/datasources/DefaultDatasourcesSpec.scala (hunk 1 above)
Comment (wemrysi) [reply impl/src/test/scala/quasar/impl/datasources/DefaultDatasourcesSpec.scala:333]: I think I was just using what had been used elsewhere in these tests, in this case we have `List[String]` so `===` should be fine.

Comment (alissapajer) [reply impl/src/test/scala/quasar/impl/datasources/DefaultDatasourcesSpec.scala:333]: Ah I see now. `>>*` is parameterized on `A: AsResult` and there's an [`AsResult[Boolean]`](https://github.com/etorreborre/specs2/blob/master/common/shared/src/main/scala/org/specs2/execute/AsResult.scala#L16-L18) that we have implicitly here. So booleans are fair game for result types.

**Comment by alissapajer** [comment]
And the build as built! 🎉

**Pull request merged**